# Project: Wrangling and Analyze Data

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import requests
import json

In [3]:
twitter = pd.read_csv('twitter-archive-enhanced.csv')
twitter.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

In [4]:
twitter.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [5]:
#twitter_noreply = twitter[(twitter['in_reply_to_status_id'].isnull() & (twitter['in_reply_to_user_id'].isnull()))]


In [6]:
#twitter_noreply.info()

In [7]:
twitter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [8]:
twitter.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [9]:
twitter.duplicated().sum()

0

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [10]:
image = pd.read_csv('image-predictions.tsv', sep = '\t')
image.head(10)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5  666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6  666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7  666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8  666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9  666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   
5        1    Bernese_mountain_dog  0.651137    True    English_springer   
6        1              box_turtle  0.933012   False          mud_turtle   
7        1                    chow  0.692517    True     Tibetan_mastiff   
8        1           shopping_cart  0.962465   False     shopping_basket   
9        1        miniature_poodle  0.201493    True            komondor   

    p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0  0.156665    True            Shetland_sheepdog  0.061428    True  
1  0.074192    True          Rhodesian_ridgeback  0.072010    True  
2  0.138584    True                   bloodhound  0.116197    True  
3  0.360687    True           miniature_pinscher  0.222752    True  
4  0.243682    True                     Doberman  0.154629    True  
5  0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True  
6  0.045885   False                     terrapin  0.017885   False  
7  0.058279    True                     fur_coat  0.054449   False  
8  0.014594   False             golden_retriever  0.007959    True  
9  0.192305    True  soft-coated_wheaten_terrier  0.082086    True

In [11]:
image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [12]:
List = []
with open('tweet-json.txt') as file:
    for line in file:
        List.append(json.loads(line))
        
txt = pd.DataFrame(List)
txt.head(2)

contributors coordinates                      created_at display_text_range  \
0         None        None  Tue Aug 01 16:23:56 +0000 2017            [0, 85]   
1         None        None  Tue Aug 01 00:17:27 +0000 2017           [0, 138]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 892420639486877696, 'id_str'...           39467   
1  {'media': [{'id': 892177413194625024, 'id_str'...           33819   

   favorited                                          full_text   geo  \
0      False  This is Phineas. He's a mystical boy. Only eve...  None   
1      False  This is Tilly. She's just checking pup on you....  None   

                         ...                          \
0                        ...                           
1                        ...                           

   possibly_sensitive_appealable quoted_status quoted_status_id  \
0                          False           NaN              NaN   
1                          False           NaN              NaN   

   quoted_status_id_str retweet_count  retweeted retweeted_status  \
0                   NaN          8853      False              NaN   
1                   NaN          6514      False              NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[2 rows x 31 columns]

In [13]:
txt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
contributors                     0 non-null object
coordinates                      0 non-null object
created_at                       2354 non-null object
display_text_range               2354 non-null object
entities                         2354 non-null object
extended_entities                2073 non-null object
favorite_count                   2354 non-null int64
favorited                        2354 non-null bool
full_text                        2354 non-null object
geo                              0 non-null object
id                               2354 non-null int64
id_str                           2354 non-null object
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null object
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 non-null obj

In [14]:
txt.contributors.isnull().sum()

2354

In [15]:
all_columns = pd.Series(list(twitter) + list(image) + list(txt))
all_columns[all_columns.duplicated()]

17                 tweet_id
42    in_reply_to_status_id
44      in_reply_to_user_id
57                   source
dtype: object

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



### Quality issues

Under (twitter table)

1. Remove retweets

2. Delete column 2, 3, 7, 8, 9 

3. Convert time stamp from object to datetime

4. Filter out rating_numerator less than 11

5. Any rating that is not equal to 10 should be taken away

Under (Txt table)

6. Change column 'id' to 'tweet_id'

7. Delete columns; 'contributors', 'cordinates', 'geo', all'in_reply' columns

8. Convert column 'tweet_id' from object to int64  

 

### Tidiness issues

1. Under twitter table, there is a need for the last four columns to be collapsed into one column

2. Use tweet_id to bring all three tables to one since that is the common item between them

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [16]:
# Make copies of original pieces of data
twitter_clean = twitter.copy()
image_clean = image.copy()

### QUALITY

### Issue #1:

#### Define: Delete retweets in all rows by filtering out the null values in retweeted columns

#### Code

In [17]:
twitter_clean = twitter[twitter.retweeted_status_id.isnull()]

#### Test

In [18]:
twitter_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

### Issue #2:

#### Define: Delete the following columns; in_reply_to_status_id, in_reply_to_user_id,retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp

#### Code

In [19]:
twitter_clean.drop(['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp','in_reply_to_status_id','in_reply_to_user_id'], axis=1, inplace = True)

/opt/conda/lib/python3.6/site-packages/pandas/core/frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


#### Test

In [20]:
twitter_clean

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
5     891087950875897856  2017-07-29 00:08:17 +0000   
6     890971913173991426  2017-07-28 16:27:12 +0000   
7     890729181411237888  2017-07-28 00:22:40 +0000   
8     890609185150312448  2017-07-27 16:25:51 +0000   
9     890240255349198849  2017-07-26 15:59:51 +0000   
10    890006608113172480  2017-07-26 00:31:25 +0000   
11    889880896479866881  2017-07-25 16:11:53 +0000   
12    889665388333682689  2017-07-25 01:55:32 +0000   
13    889638837579907072  2017-07-25 00:10:02 +0000   
14    889531135344209921  2017-07-24 17:02:04 +0000   
15    889278841981685760  2017-07-24 00:19:32 +0000   
16    888917238123831296  2017-07-23 00:22:39 +0000   
17    888804989199671297  2017-07-22 16:56:37 +0000   
18    888554962724278272  2017-07-22 00:23:06 +0000   
20    888078434458587136  2017-07-20 16:49:33 +0000   
21    887705289381826560  2017-07-19 16:06:48 +0000   
22    887517139158093824  2017-07-19 03:39:09 +0000   
23    887473957103951883  2017-07-19 00:47:34 +0000   
24    887343217045368832  2017-07-18 16:08:03 +0000   
25    887101392804085760  2017-07-18 00:07:08 +0000   
26    886983233522544640  2017-07-17 16:17:36 +0000   
27    886736880519319552  2017-07-16 23:58:41 +0000   
28    886680336477933568  2017-07-16 20:14:00 +0000   
29    886366144734445568  2017-07-15 23:25:31 +0000   
30    886267009285017600  2017-07-15 16:51:35 +0000   
...                  ...                        ...   
2326  666411507551481857  2015-11-17 00:24:19 +0000   
2327  666407126856765440  2015-11-17 00:06:54 +0000   
2328  666396247373291520  2015-11-16 23:23:41 +0000   
2329  666373753744588802  2015-11-16 21:54:18 +0000   
2330  666362758909284353  2015-11-16 21:10:36 +0000   
2331  666353288456101888  2015-11-16 20:32:58 +0000   
2332  666345417576210432  2015-11-16 20:01:42 +0000   
2333  666337882303524864  2015-11-16 19:31:45 +0000   
2334  666293911632134144  2015-11-16 16:37:02 +0000   
2335  666287406224695296  2015-11-16 16:11:11 +0000   
2336  666273097616637952  2015-11-16 15:14:19 +0000   
2337  666268910803644416  2015-11-16 14:57:41 +0000   
2338  666104133288665088  2015-11-16 04:02:55 +0000   
2339  666102155909144576  2015-11-16 03:55:04 +0000   
2340  666099513787052032  2015-11-16 03:44:34 +0000   
2341  666094000022159362  2015-11-16 03:22:39 +0000   
2342  666082916733198337  2015-11-16 02:38:37 +0000   
2343  666073100786774016  2015-11-16 01:59:36 +0000   
2344  666071193221509120  2015-11-16 01:52:02 +0000   
2345  666063827256086533  2015-11-16 01:22:45 +0000   
2346  666058600524156928  2015-11-16 01:01:59 +0000   
2347  666057090499244032  2015-11-16 00:55:59 +0000   
2348  666055525042405380  2015-11-16 00:49:46 +0000   
2349  666051853826850816  2015-11-16 00:35:11 +0000   
2350  666050758794694657  2015-11-16 00:30:50 +0000   
2351  666049248165822465  2015-11-16 00:24:50 +0000   
2352  666044226329800704  2015-11-16 00:04:52 +0000   
2353  666033412701032449  2015-11-15 23:21:54 +0000   
2354  666029285002620928  2015-11-15 23:05:30 +0000   
2355  666020888022790149  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
5     <a href="http://twitter.com/download/iphone" r...   
6     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
8     <a href="http://twitter.com/download/iphone" r...   
9     <a href="

### Issue #3:

#### Define: Convert timestamp from object to datetime 

#### Code

In [21]:
twitter_clean['timestamp'] = pd.to_datetime(twitter_clean['timestamp'])

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


#### Test

In [22]:
twitter_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2175 non-null int64
timestamp             2175 non-null datetime64[ns]
source                2175 non-null object
text                  2175 non-null object
expanded_urls         2117 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: datetime64[ns](1), int64(3), object(8)
memory usage: 220.9+ KB


### Issue #4: 

#### Define: Filter out rating_numerator less than 11

#### Code

In [23]:
twitter_clean = twitter_clean[twitter_clean.rating_numerator >= 11]


#### Test

In [24]:
twitter_clean.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

### Issue #5:

#### Define: Filter out rating_denominator not equal 10

#### Code

In [25]:
twitter_clean = twitter_clean[twitter.rating_denominator== 10]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


#### Test

In [26]:
twitter_clean.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

### Issue #6:

#### Define: Change column 'id' to 'tweet_id' in txt table

#### Code

In [27]:
txt_clean = txt.copy()

In [28]:
txt_clean.rename({'id_str': 'tweet_id'}, axis=1, inplace=True)

#### Test

In [29]:
txt_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
contributors                     0 non-null object
coordinates                      0 non-null object
created_at                       2354 non-null object
display_text_range               2354 non-null object
entities                         2354 non-null object
extended_entities                2073 non-null object
favorite_count                   2354 non-null int64
favorited                        2354 non-null bool
full_text                        2354 non-null object
geo                              0 non-null object
id                               2354 non-null int64
tweet_id                         2354 non-null object
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null object
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 non-null obj

### Issue #7:

#### Define: Delete columns; 'contributors', 'cordinates', 'geo', all'in_reply' columns

#### Code:

In [30]:
txt_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
contributors                     0 non-null object
coordinates                      0 non-null object
created_at                       2354 non-null object
display_text_range               2354 non-null object
entities                         2354 non-null object
extended_entities                2073 non-null object
favorite_count                   2354 non-null int64
favorited                        2354 non-null bool
full_text                        2354 non-null object
geo                              0 non-null object
id                               2354 non-null int64
tweet_id                         2354 non-null object
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null object
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 non-null obj

In [31]:
#txt_clean.drop(['contributors', 'coordinates', 'geo', 'place'], axis = 1, inplace = True)

#txt_clean.drop(['in_reply_to_screen_name', 'in_reply_to_status_id', 'in_reply_to_user_id', 'in_reply_to_user_id_str', 'quoted_status', 'quoted_status_id', 'quoted_status_id_str', 'retweeted_status'], axis = 1, inplace = True)

#txt_clean.drop(['in_reply_to_status_id_str'], axis = 1, inplace = True)

#### Test

In [32]:
txt_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
contributors                     0 non-null object
coordinates                      0 non-null object
created_at                       2354 non-null object
display_text_range               2354 non-null object
entities                         2354 non-null object
extended_entities                2073 non-null object
favorite_count                   2354 non-null int64
favorited                        2354 non-null bool
full_text                        2354 non-null object
geo                              0 non-null object
id                               2354 non-null int64
tweet_id                         2354 non-null object
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null object
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 non-null obj

### Issue #8:

#### Define: Convert column 'tweet_id' from object to int64

#### Code:

In [33]:
 txt_clean['tweet_id'] = txt_clean['tweet_id'].astype('int')

#### Test:

In [34]:
txt_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
contributors                     0 non-null object
coordinates                      0 non-null object
created_at                       2354 non-null object
display_text_range               2354 non-null object
entities                         2354 non-null object
extended_entities                2073 non-null object
favorite_count                   2354 non-null int64
favorited                        2354 non-null bool
full_text                        2354 non-null object
geo                              0 non-null object
id                               2354 non-null int64
tweet_id                         2354 non-null int64
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null object
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 non-null obje

### TIDINESS

### Issue #1:

#### Define: Collapse the last four columns in twitter_clean table into one column

#### Code:

In [35]:

twitter_clean['dog_types'] = twitter_clean['text'].str.extract('(doggo|floofer|pupper|puppo)')


In [36]:
twitter_clean = twitter_clean.drop(['doggo','floofer', 'pupper', 'puppo'], axis = 1)

#### Test:

In [37]:
#twitter_clean.info()
twitter_clean.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name dog_types  
0                  10   Phineas       NaN  
1                  10     Tilly       NaN  
2                  10    Archie       NaN  
3                  10     Darla       NaN  
4                  10  Franklin       NaN

### Issue #2:

#### Define: Merge all three tables(

#### Code:

In [38]:
twitter_clean = pd.merge(twitter_clean, image_clean, on = ['tweet_id'], how = 'left')

In [39]:
twitter_archive_master = pd.merge(twitter_clean, txt_clean, on = ['tweet_id'], how = 'left')

#### Test:

In [40]:
twitter_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1285 entries, 0 to 1284
Data columns (total 50 columns):
tweet_id                         1285 non-null int64
timestamp                        1285 non-null datetime64[ns]
source_x                         1285 non-null object
text                             1285 non-null object
expanded_urls                    1246 non-null object
rating_numerator                 1285 non-null int64
rating_denominator               1285 non-null int64
name                             1285 non-null object
dog_types                        240 non-null object
jpg_url                          1147 non-null object
img_num                          1147 non-null float64
p1                               1147 non-null object
p1_conf                          1147 non-null float64
p1_dog                           1147 non-null object
p2                               1147 non-null object
p2_conf                          1147 non-null float64
p2_dog                

## FINAL CLEANING

#### Drop Some Columns where necessary

In [41]:
#twitter_archive_master.drop(['contributors', 'coordinates', 'geo', 'in_reply_to_screen_name'], axis = 1, inplace = True)
#twitter_archive_master.drop(['in_reply_to_status_id', 'in_reply_to_status_id_str', 'in_reply_to_user_id', 'in_reply_to_user_id_str', 'place', 'retweeted_status'], axis = 1, inplace = True)
twitter_archive_master.drop(['quoted_status', 'quoted_status_id', 'quoted_status_id_str'], axis = 1, inplace = True)


In [42]:
twitter_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1285 entries, 0 to 1284
Data columns (total 47 columns):
tweet_id                         1285 non-null int64
timestamp                        1285 non-null datetime64[ns]
source_x                         1285 non-null object
text                             1285 non-null object
expanded_urls                    1246 non-null object
rating_numerator                 1285 non-null int64
rating_denominator               1285 non-null int64
name                             1285 non-null object
dog_types                        240 non-null object
jpg_url                          1147 non-null object
img_num                          1147 non-null float64
p1                               1147 non-null object
p1_conf                          1147 non-null float64
p1_dog                           1147 non-null object
p2                               1147 non-null object
p2_conf                          1147 non-null float64
p2_dog                

In [43]:
twitter_archive_master.head(20)

tweet_id           timestamp  \
0   892420643555336193 2017-08-01 16:23:56   
1   892177421306343426 2017-08-01 00:17:27   
2   891815181378084864 2017-07-31 00:18:03   
3   891689557279858688 2017-07-30 15:58:51   
4   891327558926688256 2017-07-29 16:00:24   
5   891087950875897856 2017-07-29 00:08:17   
6   890971913173991426 2017-07-28 16:27:12   
7   890729181411237888 2017-07-28 00:22:40   
8   890609185150312448 2017-07-27 16:25:51   
9   890240255349198849 2017-07-26 15:59:51   
10  890006608113172480 2017-07-26 00:31:25   
11  889880896479866881 2017-07-25 16:11:53   
12  889665388333682689 2017-07-25 01:55:32   
13  889638837579907072 2017-07-25 00:10:02   
14  889531135344209921 2017-07-24 17:02:04   
15  889278841981685760 2017-07-24 00:19:32   
16  888917238123831296 2017-07-23 00:22:39   
17  888804989199671297 2017-07-22 16:56:37   
18  888554962724278272 2017-07-22 00:23:06   
19  888078434458587136 2017-07-20 16:49:33   

                                             source_x  \
0   <a href="http://twitter.com/download/iphone" r...   
1   <a href="http://twitter.com/download/iphone" r...   
2   <a href="http://twitter.com/download/iphone" r...   
3   <a href="http://twitter.com/download/iphone" r...   
4   <a href="http://twitter.com/download/iphone" r...   
5   <a href="http://twitter.com/download/iphone" r...   
6   <a href="http://twitter.com/download/iphone" r...   
7   <a href="http://twitter.com/download/iphone" r...   
8   <a href="http://twitter.com/download/iphone" r...   
9   <a href="http://twitter.com/download/iphone" r...   
10  <a href="http://twitter.com/download/iphone" r...   
11  <a href="http://twitter.com/download/iphone" r...   
12  <a href="http://twitter.com/download/iphone" r...   
13  <a href="http://twitter.com/download/iphone" r...   
14  <a href="http://twitter.com/download/iphone" r...   
15  <a href="http://twitter.com/download/iphone" r...   
16  <a href="http://twitter.com/download/iphone" r...   
17  <a href="http://twitter.com/download/iphone" r...   
18  <a href="http://twitter.com/download/iphone" r...   
19  <a href="http://twitter.com/download/iphone" r...   

                                                 text  \
0   This is Phineas. He's a mystical boy. Only eve...   
1   This is Tilly. She's just checking pup on you....   
2   This is Archie. He is a rare Norwegian Pouncin...   
3   This is Darla. She commenced a snooze mid meal...   
4   This is Franklin. He would like you to stop ca...   
5   Here we have a majestic great white breaching ...   
6   Meet Jax. He enjoys ice cream so much he gets ...   
7   When you watch your owner call another dog a g...   
8   This is Zoey. She doesn't want to be one of th...   
9   This is Cassie. She is a college pup. Studying...   
10  This is Koda. He is a South Australian decksha...   
11  This is Bruno. He is a service shark. Only get...   
12  Here's a puppo that seems to be on the fence a...   
13  This is Ted. He does his best. Sometimes that'...   
14  This is Stuart. He's sporting his favorite fan...   
15  This is Oliver. You're witnessing one of his m...   
16  This is Jim. He found a fren. Taught him how t...   
17  This is Zeke. He has a new stick. Very proud o...   
18  This is Ralphus. He's powering up. Attempting ...   
19  This is Gerald. He was just told he didn't get...   

                                        expanded_urls  rating_numerator  \
0   https://twitter.com/dog_rates/status/892420643...                13   
1   https://twitter.com/dog_rates/status/892177421...                13   
2   https://twitter.com/dog_rates/status/891815181...                12   
3   https://twitter.com/dog_rates/status/891689557...                13   
4   https://twitter.com/dog_rates/status/891327558...                12   
5   https://twitter.com/dog_rates/status/891087950...                13   
6   https://gofundme.com/ydvmve-surgery-for-jax,ht...                13   
7   https://twitter.com/dog_rates/status/890729181.

In [44]:
twitter_archive_master.describe()

tweet_id  rating_numerator  rating_denominator      img_num  \
count  1.285000e+03       1285.000000              1285.0  1147.000000   
mean   7.653579e+17         14.701946                10.0     1.283348   
std    6.964010e+16         55.092579                 0.0     0.659811   
min    6.661022e+17         11.000000                10.0     1.000000   
25%    6.974829e+17         11.000000                10.0     1.000000   
50%    7.648575e+17         12.000000                10.0     1.000000   
75%    8.255351e+17         13.000000                10.0     1.000000   
max    8.924206e+17       1776.000000                10.0     4.000000   

           p1_conf      p2_conf       p3_conf  favorite_count            id  \
count  1147.000000  1147.000000  1.147000e+03     1285.000000  1.285000e+03   
mean      0.618273     0.134356  5.888854e-02    12391.653696  7.653579e+17   
std       0.267157     0.101338  5.204171e-02    13992.729601  6.964010e+16   
min       0.059033     0.000010  5.595040e-07       52.000000  6.661022e+17   
25%       0.384064     0.049964  1.268455e-02     3290.000000  6.974829e+17   
50%       0.625129     0.117397  4.643950e-02     8100.000000  7.648575e+17   
75%       0.869068     0.196623  9.332520e-02    16983.000000  8.255351e+17   
max       0.999984     0.467678  2.710420e-01   132810.000000  8.924206e+17   

       in_reply_to_status_id  in_reply_to_user_id  retweet_count  
count           5.600000e+01         5.600000e+01    1285.000000  
mean            7.551644e+17         1.304473e+16    3793.587549  
std             7.673706e+16         9.761784e+16    5520.591826  
min             6.658147e+17         1.185634e+07       0.000000  
25%             6.756548e+17         1.909316e+08    1038.000000  
50%             7.430303e+17         4.196984e+09    2268.000000  
75%             8.334487e+17         4.196984e+09    4429.000000  
max             8.862664e+17         7.305050e+17   79515.000000

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

In [ ]:
twitter_archive_master.to_csv('twitter_archive_master.csv', index=False)

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

### Insights:

1. Pupper dog type has the highest value count of 139 while floofer has the lowest value count of 4

2. The mean of rating_numerator falls between 14 - 15

3. The highest retweet_count is 79515 which has a tweet_id of 892420643555336193

### Visualization

In [ ]:
ax = twitter_archive_master.plot.hist(x='dog_types', y='favorite_count')